File: linked_list/doubly_linked_list.go
// Source: doubly_linked_list.go
// Compare this snippet from linked_list/circular_linked_list.go: now every node has a previous pointer
package main

import (
	"fmt"
)

type DoublyLinkedListNode[DoublyLinkedListNodeType comparable] struct {
	data     DoublyLinkedListNodeType
	next     *DoublyLinkedListNode[DoublyLinkedListNodeType]
	previous *DoublyLinkedListNode[DoublyLinkedListNodeType]
}

type DoublyLinkedList[DoublyLinkedListNodeType comparable] struct {
	head *DoublyLinkedListNode[DoublyLinkedListNodeType]
	size uint32
}

// Creates a new doubly linked list with only one initial node
func NewDoublyLinkedList[DoublyLinkedListNodeType comparable](data DoublyLinkedListNodeType) *DoublyLinkedList[DoublyLinkedListNodeType] {
	initialNode := &DoublyLinkedListNode[DoublyLinkedListNodeType]{data: data}
	return &DoublyLinkedList[DoublyLinkedListNodeType]{head: initialNode, size: 1}
}

// Inserts a new node next to the head node
func (dll *DoublyLinkedList[DoublyLinkedListNodeType]) InsertNextToHead(data DoublyLinkedListNodeType) {
	// ... <-> head <-> newNode <-> ...
	if dll.head == nil {
		fmt.Println("The head of the doubly linked list is nil")
		return
	}

	newNode := &DoublyLinkedListNode[DoublyLinkedListNodeType]{
		data:     data,
		next:     dll.head.next,
		previous: dll.head,
	}

	if dll.head.next != nil {
		dll.head.next.previous = newNode
	}

	dll.head.next = newNode
	dll.size++
}

// Inserts a new node before the head node
func (dll *DoublyLinkedList[DoublyLinkedListNodeType]) InsertBeforeHead(data DoublyLinkedListNodeType) {
	// ... <-> newNode <-> head <-> ...
	newNode := &DoublyLinkedListNode[DoublyLinkedListNodeType]{
		data:     data,
		next:     dll.head,
		previous: nil,
	}

	if dll.head != nil {
		dll.head.previous = newNode
	}

	dll.head = newNode
	dll.size++
}

// Inserts a new node after a node with a specific value
func (dll *DoublyLinkedList[DoublyLinkedListNodeType]) InsertAfterNodeByValue(data DoublyLinkedListNodeType, value DoublyLinkedListNodeType) {
	currentNode := dll.head
	for currentNode != nil {
		if currentNode.data == value {
			// Match found, insert the new node after the current node
			// ... <-> currentNode <-> newNode <-> ...
			newNode := &DoublyLinkedListNode[DoublyLinkedListNodeType]{
				data:     data,
				next:     currentNode.next,
				previous: currentNode,
			}

			// If the current node is not the last node, connect the next node to the new node
			if currentNode.next != nil {
				currentNode.next.previous = newNode
			}

			currentNode.next = newNode
			dll.size++
			return
		}
		currentNode = currentNode.next
	}

	fmt.Printf("Node with value %v not found\n", value)
}

// Deletes a node with a specific value
func (dll *DoublyLinkedList[DoublyLinkedListNodeType]) DeleteNodeByValue(value DoublyLinkedListNodeType) {
	currentNode := dll.head
	for currentNode != nil {
		if currentNode.data == value {
			// Match found, delete the current node
			// ... <-> previousNode <-> currentNode <-> nextNode <-> ...
			if currentNode.previous != nil {
				currentNode.previous.next = currentNode.next
			}

			// If the current node is not the last node, connect the previous node to the next node
			if currentNode.next != nil {
				currentNode.next.previous = currentNode.previous
			}

			dll.size--
			return
		}
		currentNode = currentNode.next
	}

	fmt.Printf("Node with value %v not found\n", value)
}

// Displays the doubly linked list
func (dll *DoublyLinkedList[DoublyLinkedListNodeType]) Display() {
	if dll.size == 0 {
		fmt.Println("The doubly linked list is empty")
		return
	}
	currentNode := dll.head
	for currentNode.next != nil {
		fmt.Printf("%v <-> ", currentNode.data)
		currentNode = currentNode.next
	}

	// Emphasize circularity
	fmt.Printf("%v\n", dll.head.data)
}

func useDoublyLinkedList() {
	dll := NewDoublyLinkedList(1)
	dll.InsertNextToHead(2)
	dll.Display()

	dll.InsertBeforeHead(3)
	dll.Display()

	dll.InsertBeforeHead(4)
	dll.Display()

	dll.InsertAfterNodeByValue(5, 3)
	dll.Display()

	dll.InsertAfterNodeByValue(6, 2)
	dll.Display()

	dll.DeleteNodeByValue(3)
	dll.Display()
}

// 1 <-> 1
// 3 <-> 1 <-> 3
// 4 <-> 3 <-> 1 <-> 4
// 4 <-> 3 <-> 5 <-> 1 <-> 4
// 4 <-> 3 <-> 5 <-> 1 <-> 2 <-> 4
// 4 <-> 5 <-> 1 <-> 2 <-> 4

File: linked_list/doubly_linked_list.c
/**
 * @file doubly_linked_list.c
 * @brief A simple implementation of doubly linked list, written in C langauge.
 */
//

#include <stdio.h>
#include <stdlib.h>
#include <memory.h>
#include <string.h>
#include <stdbool.h>

typedef char DoublyLinkedListElement;
typedef struct DoublyLinkedListNode {
    DoublyLinkedListElement data;
    struct DoublyLinkedListNode *next;
    struct DoublyLinkedListNode *previous;
} DoublyLinkedListNode;

// Initialize a new doubly linked list node
void initializeDoublyLinkedListNode(DoublyLinkedListNode *node, 
                                    DoublyLinkedListElement data) {
    // // Fill the memory with 0 (initialize the node with 0 to ensure safety
    memset(node, 0, sizeof(DoublyLinkedListNode));
    node->data = data;
    node->next = node;
    node->previous = node;
}

// Insert a new node next to the given node
void insertDoublyLinkedListNodeNextTo(DoublyLinkedListNode *node, 
                                      DoublyLinkedListElement data) {
    DoublyLinkedListNode *newNode = (DoublyLinkedListNode *)malloc(sizeof(DoublyLinkedListNode));
    initializeDoublyLinkedListNode(newNode, data);

    // ... <-> node <-> newNode <-> ...
    newNode->next = node->next;         // newNode->next points to the node that was next to the given node (forward)
    newNode->previous = node;           // newNode->previous points to the given node (backward)
    node->next->previous = newNode;     // The node that was next to the given node now points to the newNode backward
    node->next = newNode;               // The given node now points to the newNode forward
}

// Insert a new node next to a node that has a specific data
void insertDoublyLinkedListNodeNextToData(DoublyLinkedListNode *headNode,
                                          DoublyLinkedListElement data,
                                          DoublyLinkedListElement newData) {
    DoublyLinkedListNode *currentNode = headNode;

    do {
        if (currentNode->data == data) {
            insertDoublyLinkedListNodeNextTo(currentNode, newData);
            return;
        }
    } while ((currentNode = currentNode->next) != headNode);

    fprintf(stderr, "The data %c does not exist in the doubly linked list.\n", data);
}

// Remove a node from the doubly linked list
void removeDoublyLinkedListNode(DoublyLinkedListNode *node) {
    // ... <-> previousNode <-> node <-> nextNode <-> ...
    node->previous->next = node->next;               // The node before the given node now points to the node after the given node
    node->next->previous = node->previous;           // The node after the given node now points to the node before the given node
    free(node);                               // Free the memory of the given node
}

// Print the doubly linked list
void printDoublyLinkedList(DoublyLinkedListNode *headNode) {
    if (headNode == NULL) {
        printf("The doubly linked list is empty.\n");
        return;
    }

    DoublyLinkedListNode *currentNode = headNode;
    do {
        printf("%c <-> ", currentNode->data);
        currentNode = currentNode->next;
    } while (currentNode != headNode);

    // Emphasize the circularity of the doubly linked list
    printf("%c\n", headNode->data);
}

// Free the memory of the doubly linked list
void freeDoublyLinkedList(DoublyLinkedListNode *headNode) {
    if (headNode == NULL) {
        printf("The doubly linked list is empty.\n");
        return;
    }

    DoublyLinkedListNode *currentNode = headNode;
    do {
        DoublyLinkedListNode *tempNode = currentNode;
        currentNode = currentNode->next;
        free(tempNode);
    } while (currentNode != headNode);

    free(headNode);
    printf("The doubly linked list has been freed.\n");
}

int main(void) {
    // Initialize a new doubly linked list
    DoublyLinkedListNode *doublyLinkedList = (DoublyLinkedListNode *)malloc(sizeof(DoublyLinkedListNode));

    // Initialize the head node of the doubly linked list
    initializeDoublyLinkedListNode(doublyLinkedList, 'A');
    printDoublyLinkedList(doublyLinkedList);

    // Inserting
    insertDoublyLinkedListNodeNextTo(doublyLinkedList, 'B');
    printDoublyLinkedList(doublyLinkedList);
    insertDoublyLinkedListNodeNextTo(doublyLinkedList->next, 'C');
    printDoublyLinkedList(doublyLinkedList);
    insertDoublyLinkedListNodeNextTo(doublyLinkedList->next->next, 'D');
    printDoublyLinkedList(doublyLinkedList);
    insertDoublyLinkedListNodeNextTo(doublyLinkedList->next->next->next, 'E');
    printDoublyLinkedList(doublyLinkedList);

    // Inserting next to a node that has a specific data
    insertDoublyLinkedListNodeNextToData(doublyLinkedList, 'B', 'X');
    printDoublyLinkedList(doublyLinkedList);
    insertDoublyLinkedListNodeNextToData(doublyLinkedList, 'C', 'Y');
    printDoublyLinkedList(doublyLinkedList);

    // Removing
    removeDoublyLinkedListNode(doublyLinkedList->next->next);
    printDoublyLinkedList(doublyLinkedList);
    removeDoublyLinkedListNode(doublyLinkedList->next);
    printDoublyLinkedList(doublyLinkedList);

    freeDoublyLinkedList(doublyLinkedList);

    return 0;
}

// A <-> A
// A <-> B <-> A
// A <-> B <-> C <-> A
// A <-> B <-> C <-> D <-> A
// A <-> B <-> C <-> D <-> E <-> A
// A <-> B <-> X <-> C <-> D <-> E <-> A
// A <-> B <-> X <-> C <-> Y <-> D <-> E <-> A
// A <-> B <-> C <-> Y <-> D <-> E <-> A
// A <-> C <-> Y <-> D <-> E <-> A